app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,229 @@
+        Log.d(TAG, activity.getString(R.string.discovery_failed) + ": " + errorMessage);
+        showToast(activity,
+            activity.getString(R.string.discovery_failed),
+            Toast.LENGTH_SHORT);
+      }
+    });
+  }
+
+  public void setWifiP2pEnabled(boolean wifiP2pEnabled) {
+    this.wifiP2pEnabled = wifiP2pEnabled;
+  }
+
+  public boolean isWifiP2pEnabled() {
+    return wifiP2pEnabled;
+  }
+
+  /* From WifiP2pManager.ChannelListener interface */
+  @Override
+  public void onChannelDisconnected() {
+    // Upon disconnection, retry one more time
+    if (manager != null && !retryChannel) {
+      Log.d(TAG, "Channel lost, trying again");
+      activity.clearPeers();
+      retryChannel = true;
+      manager.initialize(activity, getMainLooper(), this);
+    } else {
+      showToast(activity, R.string.severe_loss_error, Toast.LENGTH_LONG);
+    }
+  }
+
+  /* From WifiP2pManager.PeerListListener callback-interface */
+  @Override
+  public void onPeersAvailable(@NonNull WifiP2pDeviceList peers) {
+    ((Callbacks) activity).updatePeerDevicesList(peers);
+  }
+
+  /* From WifiP2pManager.ConnectionInfoListener callback-interface */
+  @Override
+  public void onConnectionInfoAvailable(@NonNull WifiP2pInfo groupInfo) {
+    /* Devices have successfully connected, and 'info' holds information about the wifi p2p group formed */
+    setGroupInfo(groupInfo);
+    ((Callbacks) activity).performHandshakeWithSelectedPeerDevice(groupInfo);
+  }
+
+  public void setUserDevice(@NonNull WifiP2pDevice userDevice) {
+    this.userDevice = userDevice;
+  }
+
+  public void setGroupInfo(@NonNull WifiP2pInfo groupInfo) {
+    this.groupInfo = groupInfo;
+  }
+
+  public boolean isGroupFormed() {
+    return groupInfo.groupFormed;
+  }
+
+  public boolean isGroupOwner() {
+    return groupInfo.isGroupOwner;
+  }
+
+  public @NonNull InetAddress getGroupOwnerAddress() {
+    return groupInfo.groupOwnerAddress;
+  }
+
+  public void setSenderSelectedPeerDevice(@NonNull WifiP2pDevice senderSelectedPeerDevice) {
+    this.senderSelectedPeerDevice = senderSelectedPeerDevice;
+  }
+
+  public void connect() {
+    if(senderSelectedPeerDevice == null) {
+      Log.d(TAG, "No device set as selected");
+    }
+
+    WifiP2pConfig config = new WifiP2pConfig();
+    config.deviceAddress = senderSelectedPeerDevice.deviceAddress;
+    config.wps.setup = WpsInfo.PBC;
+
+    manager.connect(channel, config, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        // UI updated from broadcast receiver
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.connection_failed) + ": " + errorMessage);
+        showToast(activity, activity.getString(R.string.connection_failed),
+            Toast.LENGTH_LONG);
+      }
+    });
+  }
+
+  public void closeLocalFileTransferActivity() {
```

Comment:
not the repsosibility of WifiDirectManager, it should on a fatal error or success clean itself up and then report this via callback and let the activity decide

Reply:
I'm only calling finish() now, and destroyWifiDirectManager() handles the cleanup (called in onDestroy())
[Commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/19b316644e88a12281b92b890d156bea07572ddd)